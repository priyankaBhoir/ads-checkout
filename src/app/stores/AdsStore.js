'use strict'

import dispatcher from '../dispatcher/Dispatcher';
import {EventEmitter} from 'events';
import { createStore } from './Store';
import Constants from '../constants/Constants';

const {
	API_CONSTANT,
	EVENT_CONSTANT
} = Constants;

let _allAds = [];

const AdsStore = createStore ({
	currentState () {
		return _allAds;
	}
});

function updateAds (data) {
	_allAds = data;
}

AdsStore.dispatchToken = dispatcher.register (action => {

	switch (action.type) {
		case API_CONSTANT.LOAD_DATA:
			updateAds(action.data);
			AdsStore.emitChange(EVENT_CONSTANT.ADS_LOADED);
			break;
	}
});

export default AdsStore;