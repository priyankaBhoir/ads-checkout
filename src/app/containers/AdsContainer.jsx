import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {loadAllAds} from '../actions/AdActions';
import {addToCart, checkOut} from '../actions/CartActions';
import AdsStore from '../stores/AdsStore';
import CartStore from '../stores/CartStore';
import Constant from '../constants/Constants'
import AdShelf from "../components/AdShelf.jsx";
import {
  Link
} from 'react-router-dom';

export default class AdsContainer extends Component {
	constructor(props) {
		super(props);
		this.state = {
			ads: AdsStore.currentState(),
			cartItemNos: CartStore.getTotalNumberOfItems()
		}
	}

	componentDidMount() {
		AdsStore.addChangeListner(Constant.EVENT_CONSTANT.ADS_LOADED, this.updateAllAds)
		CartStore.addChangeListner(Constant.EVENT_CONSTANT.ITEM_ADDED_TO_CART, this.updateCartItemNos)
		loadAllAds();
	}

	componentWillUnmount() {
		AdsStore.removeChangeListner(Constant.EVENT_CONSTANT.ADS_LOADED, this.updateAllAds)
		CartStore.removeChangeListner(Constant.EVENT_CONSTANT.ITEM_ADDED_TO_CART, this.updateCartItemNos)
	}

	updateAllAds = () => {
		console.log("dgvgfd");
		this.setState({
			ads: AdsStore.currentState()
		})
	}

	updateCartItemNos = () => {
		this.setState({
			cartItemNos: CartStore.getTotalNumberOfItems()
		})
	}

	render() {
		return !this.state.ads.length ? <div> Loading ... </div> : (
			<div class="section">
				<div className="header"> <span className="btn btn-floating right" style={{marginTop: -40}}> {this.state.cartItemNos}</span></div>
				<div className="row"> {this.state.ads.map(ad => <AdShelf key={ad.id} ad={ad} addToCart={addToCart}/>)} </div>
				<div className="center"> <span class="btn grey lighten-2"> <Link to="/cart">goto cart</Link></span></div>
			</div>
		)
	}
}
