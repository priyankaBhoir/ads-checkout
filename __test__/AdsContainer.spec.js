import React from 'react'
import { shallow, mount, unmount } from 'enzyme';
import renderer from 'react-test-renderer';
import {MemoryRouter } from 'react-router-dom'
import AdsContainer from '../src/app/containers/AdsContainer';

import Dispatcher from '../src/app/dispatcher/Dispatcher';
import AppConstants from '../src/app/constants/Constants';
import AdsStore from '../src/app/stores/AdsStore';

let componentDidMount = jest.fn();
let updateAllAds = jest.fn();
AdsContainer.prototype.updateAllAds = updateAllAds;

AdsContainer.prototype.componentDidMount = jest.fn(() => {
  AdsStore.addChangeListner(AppConstants.EVENT_CONSTANT.ADS_LOADED, updateAllAds)
});


jest.mock('react-router-dom');


const mountWithRouter = node => mount(<MemoryRouter>{node}</MemoryRouter>);

// Snapshot for Home React Component
describe('AdShelf --- Snapshot',()=>{
    it('+++capturing Snapshot of Container', () => {
        const renderedValue =  renderer.create(<AdsContainer />).toJSON()
        expect(renderedValue).toMatchSnapshot();

    });
});

///*******************************************************************************************************
describe('Shallow Render REACT [SMART] Container :: AdsContainer',()=>{
    let wrapper
    beforeEach(()=>{
        if(wrapper) wrapper.unmount();
       wrapper = mountWithRouter(<AdsContainer/>);     
    })
    
    let spy

    afterEach(() => {
      spy.mockClear()
    })

    it('+++ Smart component should invoke method to load data', () => {
       
       if(wrapper) wrapper.unmount();
       spy = jest.spyOn(AdsContainer.prototype, 'componentDidMount');
        // wrapper = mount(<MemoryRouter initialEntries={[ '/' ]}  initialIndex={0}> <AdsContainer/> </MemoryRouter>);
        wrapper = mountWithRouter(<AdsContainer />)
        // expect(componentDidMount).toBeCalled();
        expect(spy).toHaveBeenCalled();
    });

   it('+++ SMART Component should listen to changes from store', () => {
        // dispatch action ( store is listening for action )
        spy = jest.spyOn(AdsContainer.prototype, 'updateAllAds');
        Dispatcher.dispatch({
            type: AppConstants.API_CONSTANT.LOAD_DATA,
            data: [{
              id: 'classic',
              title: 'Classic',
              price: 299,
              description: ["Basic Ad"]
          }]
        });
        
        expect(spy).toHaveBeenCalled();
    });
    
});
