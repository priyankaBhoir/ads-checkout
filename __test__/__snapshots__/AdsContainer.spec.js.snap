// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdShelf --- Snapshot +++capturing Snapshot of Container 1`] = `
<div>
   Loading ... 
</div>
`;
